auslib/db.py

```diff
@@ -665,9 +718,243 @@ def rollbackChange(self, change_id, changed_by, transaction=None):
+            base_table_where = {pk: new_row["base_%s" % pk] for pk in self.base_primary_key}
+            if new_row.get("base_data_version"):
+                self.baseTable.update(base_table_where, new_row, changed_by, new_row["base_data_version"], transaction=transaction, dryrun=True)
+            else:
+                self.baseTable.insert(changed_by, transaction=transaction, dryrun=True, **new_row)
+
+            conditions = {}
+            for cond in itertools.chain(*self.condition_groups):
+                if cond in new_row:
+                    conditions[cond] = new_row[cond]
+                elif row.get(cond):
+                    conditions[cond] = row[cond]
+
+            self._validateConditions(conditions)
+
+        if not dryrun:
+            renamed_what = self._prefixColumns(what)
+            renamed_what["scheduled_by"] = changed_by
+            return super(ScheduledChangeTable, self).update(where, renamed_what, changed_by, old_data_version, transaction)
+
+    def delete(self, where, changed_by=None, old_data_version=None, transaction=None, dryrun=False):
+        for row in self.select(where=where, transaction=transaction):
+            base_row = {col[5:]: row[col] for col in row if col.startswith("base_")}
+            base_table_where = {pk: row["base_%s" % pk] for pk in self.base_primary_key}
+            # TODO: What permissions *should* be required to delete a scheduled change?
+            # It seems a bit odd to be checking base table update/insert here. Maybe
+            # something broader should be required?
+            if base_row.get("base_data_version"):
+                self.baseTable.update(base_table_where, base_row, changed_by, base_row["base_data_version"], transaction=transaction, dryrun=True)
+            else:
+                self.baseTable.insert(changed_by, transaction=transaction, dryrun=True, **base_row)
+
+        if not dryrun:
+            return super(ScheduledChangeTable, self).delete(where, changed_by, old_data_version, transaction)
+
+    def enactChange(self, sc_id, enacted_by, transaction=None):
+        if not self.db.hasPermission(enacted_by, "scheduled_change", "enact", transaction=transaction):
+            raise PermissionDeniedError("%s is not allowed to enact scheduled changes", enacted_by)
+
+        sc = self.select(where=[self.sc_id == sc_id], transaction=transaction)[0]
+        what = {}
+        for col in sc:
+            if col.startswith("base_"):
+                what[col[5:]] = sc[col]
+
+        # The scheduled change is marked as complete first to avoid it being
+        # updated unnecessarily when the base table's update method calls
+        # mergeUpdate. If the base table update fails, this will get reverted
+        # when the transaction is rolled back.
+        self.update([self.sc_id == sc_id], {"complete": True}, changed_by=sc["scheduled_by"], old_data_version=sc["data_version"], transaction=transaction)
```

Comment:
In person we talked about moving the complete=True check into the select, so that the db does the work when the # of scheduled changes grows.